### ISSUE TITLE
SNI support to be enabled by default?

### ISSUE THREAD

**Issue by huima** [http-kit/http-kit#393]
If I would make a pull request which would enable the SNI support in http-kit client by default, is there any reasons why it would not be accepted? Are there any security or other reasons to keep the current config where SNI needs to be enabled by own ssl configurator.

We are using http-kit heavily and now need to configure SNI support to be enabled in every project.

**Comment by kumarshantanu** [comment]
AFAIK support for SNI requires Java 8, and making that a requirement would leave many Java 6/7 users out. Linking old issues on this subject: #187 #311 #335

**Comment by huima** [comment]
Yup. It could have a check via java interop to check that the java.vm.version system property is 1.8 ->

**Comment by jwr** [comment]
I've just spent the better part of a day debugging this problem and ended up here. I think the current situation is untenable: sites proxied via CloudFlare require SNI, as do many other places on the Internet.

I think support for older JVM versions is secondary. These days code running on Java 6 and 7 will not be able to connect to many modern APIs or sites anyway, because of 1) SNI, and 2) limited crypto policy that is enforced by default, so without manual installation of JCE unlimited policy files many sites will fail because of cipher negotiation.

I implemented the workaround suggested by @kumarshantanu (thanks!), but I think http-kit client is deficient right now and the lack of SNI is a major pitfall which deserves at least a mention in the README.

**Comment by ptaoussanis** [comment]
Hey everyone, sorry for the inactivity on this- haven't had an opportunity to spend any time on this.

Could someone kindly summarise the current situation, options, and trade-offs?
Is there a consensus on a particular action to take?

**Comment by jwr** [comment]
Well, I've outlined my rationale above — not sure what I could add to that. To rephrase and summarize:

1. SNI is required for many modern sites & APIs (it seems anything hosted behind Cloudflare requires it now).

2. Java 6/7 does not support SNI, so these apps will not work with modern sites & APIs anyway.

3. Many modern sites also require modern crypto like AES-256, which in Java 6 and 7 requires installing additional libraries to unblock the JCE unlimited crypto policy. This means that old Java 6/7 apps will not work anyway without changes to the JVM installation.

My conclusion is that while I love backwards compatibility, in this case I think it makes little sense, at least for the http-client part. Perhaps a last release that supports Java 6/7 could be made, and then we should start requiring Java 8 and support SNI.

**Comment by huima** [comment]
Totally agree with @jwr 

Java world has changed a lot along the years and as licensing models have also changed I kinda assume that there are less and less users who stay in Oracle's supported old JDKs / JREs - except for some enterprise applications, which then work inside a corporate network. For them these kind of troubles are not necessarily at all an issue, and neither would using a legacy version of http-kit... If they would be using clojure at all.

Rest of us who are building or interoperating with public facing internet / cloud services need a more modern and maintained JDK versions in any case.

**Comment by ptaoussanis** [comment]
> Well, I've outlined my rationale above — not sure what I could add to that.

Thanks Jan. Like I mentioned, what would be helpful to have now is:

1. A clear description/recap of the specific action to take that's being advocating for. (E.g. merge PR x).
2. A clear description/recap of trade-offs involved (E.g. this change will now mean that all library users need JVM >= x instead of JVM >= y? Or only users of the client, or only users importing a specific ns, etc.?)
3. A clear description/recap of alternative options if any are relevant, with trade-offs. (E.g. maintain 2 versions, separate client+server so folks who still want the old JVM but don't use the client can stick to the old JVM).
4. A confirmation that there's a consensus behind the specific action that's being advocated for, and that there's no significant open questions or concerns. (Or a clear description/recap of what those open questions or concerns are).

I haven't had (and won't have for some weeks) an opportunity to properly catch up on this and related threads, so if someone wants to help move this along- the above info would be a great way to assist!

Appreciate that there's folks strongly motivated to get something changed ASAP- but please keep in mind that a lot of folks depend on http-kit, and not everyone's needs (or environment) is the same. The above info represents a minimal set of clarity I think we should aim for before taking an action on this.

Thanks everyone!

**Comment by aiba** [comment]
Hi, I'm also affected by this issue (can't talk to an API server behind cloudflare).  I'm glad it's being looked into, but despite reading  #187 #311 and #335, I don't see exactly what the workaround is.  Earlier in this thread, @jwr said he implemented a workaround from @kumarshantanu.

Forgive me if I'm missing something obvious, but I would be very grateful if someone could explain the workaround.

**Comment by jwr** [comment]
@ptaoussanis is absolutely right. This is a non-trivial breaking change and it does need careful consideration.

Unfortunately, I am unable to dedicate time to this in the near future, so unless someone else takes the lead on this issue, we will have to make do with workarounds.

**Comment by aiba** [comment]
Maybe in the short term, we can document the workaround and make it easy for people to find it.

Currently, if you just use http-kit with a cloudflare server, you get an exception:

`javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure`

And it's not clear what to do.  I'm not sure people will find this issue, since the exception doesn't say anything about SNI.

I volunteer to write up a page, "Debugging SSLHandshakeException" with a bunch of things I've learned while debugging this issue myself, including the workaround for CloudFlare/SNI.  But where should this page go, and how do we help people find it?

The first thought that comes to mind is to create an open github issue with "handshake_failure" in the title.  When I hit this exception, one of the first things I did was look at the open issues, but I totally missed this one because it took me a whole day to figure out it was related to SNI.

**Comment by jwr** [comment]
@aiba I think this is a great idea. I had also spent about a day debugging and searching until I figured out the handshake failure is SNI-related.

**Comment by ptaoussanis** [comment]
Hi folks,

Quick update: @fpischedda has kindly submitted a PR to:

  - Allow the default client to be overridden.
  - Add a new optional SNI-capable client.
  - Document the above, and add an example to the README.

This should make the SNI support easier for folks to use, without breaking backward compatibility for older JVM versions. The default may still be changed in future, but I'd vote that this is probably sufficient for now - and a reasonable tradeoff.

Heads-up that we're just waiting for some unit tests to be added, after which I'll try cut a new beta release.

In the meantime have pushed `2.4.0-alpha5` (undocumented) in case anyone wants to take a look / provide feedback.

**Comment by aiba** [comment]
> The default may still be changed in future, but I'd vote that this is probably sufficient for now - and a reasonable tradeoff.

If this is not the default, I would just worry that anyone who is new to http-kit who tries out the library with a cloudflare server, will get `javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure` and then be stuck.

I would be strongly in favor of something like #423 to have http-kit "just work" on cloudflare servers, which seem to be popular for many https APIs.

**Comment by jwr** [comment]
It appears that the SNI fix in `04bf72b` does not work for me:

```
Retrieving http-kit/http-kit/2.4.0-alpha6/http-kit-2.4.0-alpha6.jar from clojars
#error {                                                                      
 :cause org.httpkit.client                                                    
 :via                                                                         
 [{:type clojure.lang.Compiler$CompilerException   
   :message Syntax error compiling at (org/httpkit/sni_client.clj:55:5).
   :data #:clojure.error{:phase :compile-syntax-check, :line 55, :column 5, :source org/httpkit/sni_client.clj}
   :at [clojure.lang.Compiler analyzeSeq Compiler.java 7115]}
  {:type java.lang.ClassNotFoundException      
   :message org.httpkit.client                                                
   :at [java.net.URLClassLoader findClass URLClassLoader.java 382]}]
```

I suspect it's because of the circular dependency workaround, but I'm not sure.

**Comment by ptaoussanis** [comment]
@jwr Hi Jan- is this still consistently a problem for you? Have you tried first clearing build artifacts (e.g. with `lein clean`)?

Took a quick look and it's not obvious to me why there'd be a problem - but I can dig further if you're still having issues.

Thanks,

**Comment by jwr** [comment]
@ptaoussanis Thanks for remembering. I re-checked and it is not a problem anymore. I do not know why, but a lot of things have changed since I last checked.

So, I can at least confirm that replacing `[http-kit "2.3.0"]` with `[http-kit "2.4.0-alpha6"]` results in a compilable build now.

**Comment by ptaoussanis** [comment]
@jwr Awesome, happy to hear it! Thanks for the confirmation. And apologies for the delay responding.

**Comment by st3fan** [comment]
I would like to revive this thread by asking if SNI can be enabled by default on runtimes that support it, as an internal library detail.

**Comment by borkdude** [comment]
FWIW, babashka already ships http-kit with SNI enabled by default (since its runtime is Java 11 this is not a problem). I would prefer a solution that is compatible with GraalVM. E.g.:

``` 
(defmacro with-java-8 [& body]
  (when (resolve 'javax.net.ssl.SNIHostName)
    `(do ~@body)))
```
and then
```
(with-java-8
  (def *default-client* ...))
```
Something like that would work.

**Comment by borkdude** [comment]
@ptaoussanis Just letting you know that I've seen the same question over and over in the beginners channel in Clojurians slack:

Why doesn't httpkit client work with this request?
- Can you test with the native babashka httpkit client (which has SNI enabled)?
Yes, that works.
- Have you enabled SNI in httpkit?
Oh, no, I haven't.

Happened multiple times this week.

**Comment by borkdude** [comment]
<img width="1427" alt="Screenshot 2021-05-05 at 19 33 45" src="https://user-images.githubusercontent.com/284934/117184291-d8fe0780-add8-11eb-999f-bf5952912424.png">

<img width="801" alt="Screenshot 2021-05-05 at 19 32 29" src="https://user-images.githubusercontent.com/284934/117184191-bc61cf80-add8-11eb-812c-95bda4ff59f0.png">

**Comment by aiba** [comment]
I've also had people in my company ask me why http-kit doesn't just work by default.  I am strongly in favor of enabling this by default.  I think even Java's built-in `URL.openConnection()` works this way by default.

**Comment by ptaoussanis** [comment]
Hi everyone, sorry for the delayed response - has been a bit of a crazy period for me at (and outside) work.

Just to clarify what we're talking about here- there's 3x cases to consider:

1. http-kit to support Java <8 and Java >=8. Auto enable SNI when Java >= 8, auto disable when Java <8.
2. http-kit to support Java <8 and Java >=8. Auto enable SNI when Java >= 8, provide some manual mechanism to still work with Java <8.
3. http-kit to support only Java >= 8. Auto enable SNI.

- Case 1 would be a no-brainer if possible, since it doesn't break anyone.
- Case 2 would be agreeable, since it's only a soft-break in the worst case.
- Case 3 would involve some tradeoffs that I think would need to be made clear to everyone first.

I've only had an opportunity to skim this thread, but if I understand correctly from Michiel's [comment](https://github.com/http-kit/http-kit/issues/393#issuecomment-735433361) that we believe we can successfully achieve case 1, then that'd be perfect and someone just needs to please submit a PR.

Am I understanding correctly?

**Comment by borkdude** [comment]
I think that's correct, although I would be surprised if people were still depending on Java 7 at this point. It's not officially support by Clojure itself anymore.

**Comment by aiba** [comment]
Echoing @borkdude, I would imagine anyone still using clojure with Java 7 is not in the habit of bumping any of the version numbers of their dependencies.  And if http-kit must always continue to support Java 7, that comes at a cost of additional complexity (such as cases 1 and 2 above).  And even more complexity for each new feature that requires different code paths to support Java 7.

So I would vote for case 3, http-kit versions above some # require Java >=8.  Java 7 users (if there are any) would just stick with a version of http-kit that supports Java 7.

**Comment by ptaoussanis** [comment]
Hi Aaron!

> And if http-kit must always continue to support Java 7, that comes at a cost of additional complexity (such as cases 1 and 2 above). And even more complexity for each new feature that requires different code paths to support Java 7.

There's several statements here that I think are possibly unsupported:

- I don't believe that anyone has attempted to argue that http-kit "must always" continue to support Java 7. I'm not even necessarily advocating for continued support now if the costs are non-trivial. I was asking for a clarification of what proposal is on the table since it was unclear to me.
- Cases 1 and 2 _may_ involve additional complexity, but I don't believe that anyone has attempted to quantify how significant it may be if it exists. (I suspect that we may be talking about something fairly trivial, and well-isolated here).
- There's no other features I'm aware of that are currently incurring complexity in an effort to retain support of Java 7.

So all I'm advocating for at this point is:

- A clear description of the precise proposal/s on the table.
- A clear description of any trade-offs implied by those proposal/s.

With the above info, we can jointly make an informed decision about how we want to proceed. This would include weighing the value of continued Java 7 support.

Just saying "enable SNI support by default!" is not specific enough, since it doesn't distinguish between cases 1, 2, 3 - which have different implementations and trade-offs.

I hope this position seems reasonable?

**Comment by aiba** [comment]
Yes, I actually agree with all of that.  And I think [your comment](https://github.com/http-kit/http-kit/issues/393#issuecomment-832899775) correctly understands and identifies the three cases.  And I agree it's a tradeoff that is hard to quantify.  I didn't mean to accuse cases 1 and 2 of being obviously bad (sorry if it came off that way).  I just wanted to weigh in with my 2¢ in favor of case 3, for what that's worth.

**Comment by ptaoussanis** [comment]
@aiba Thanks for clarifying, makes sense- think we're on the same page 👍 

> I just wanted to weigh in with my 2¢ in favor of case 3, for what that's worth.

That's definitely useful input- as was your important observation about potential complexity costs for cases 1 and 2.

**Next step**: input/ideas/PRs from anyone welcome on this. Otherwise I can try find an opportunity myself, but can't guarantee how soon that'd be since I'm going to continue to be pretty heavily loaded till end June.

**Comment by kipz** [comment]
Totally agree. I think it's about time. I think there are some other issues/PR's that are closely related too.

**Comment by ptaoussanis** [comment]
We're on the same page, SNI support will be enabled by default in v2.7 👍
I'll handle this myself as a last step before cutting the first v2.7 beta.

**Comment by ptaoussanis** [comment]
PR [513](https://github.com/http-kit/http-kit/pull/513) is up to update default client, merge is blocked on updating broken client tests. Assistance welcome for that, otherwise I'll try take a look myself in a few weeks.

**Comment by ptaoussanis** [comment]
Closing, will be addressed in upcoming v2.7 alpha 👍
